### PR TITLE
Tg 31 implement training code

### PR DESCRIPTION
**Pull request by rohit901**
@Dmmc123 please review.

Fixes various bugs in losses, discriminator, generator and main train loop.

* GANs are trained by first updating discriminator and then updating the generator. The current design had some errors with executing this logic, hence introduced instance functions in discriminator to get the logits. This allows us to update discriminator first and then update generator as is typically done for GANs. The changes here are mostly minimal but you can review it once.

* PyTorch BCE loss function already incorporates the negative sign and hence we should not be using negative sign in our loss calculations. Fixed this error, now the shown loss is positive and it updates correctly, earlier it was just showing large negative numbers which were not making sense.

* Updated the necessary tests and some other minor fixes.

### CODE REVIEW

File: .pylintrc [hunk 1]
```diff
@@ -2,4 +2,7 @@
 
 # List of members which are set dynamically and missed by Pylint inference
 # system, and so shouldn't trigger E1101 when accessed.
-generated-members=numpy.*, torch.*
\ No newline at end of file
+generated-members=numpy.*, torch.*
+
+[MESSAGES CONTROL]
+disable=not-callable
```
Comment (Dmmc123) [inline .pylintrc:8]: What is the reason for bypassing this check?
I hardly believe it was necessary and there are ways to make it work without hardcoding an explicit bypass

File: src/models/modules/discriminator.py [hunk 2]
```diff
@@ -128,29 +128,50 @@ def __init__(self) -> None:
...
+        Compute logits for conditional adversarial loss
+
+        :param torch.Tensor visual_features: Features from Inception encoder. Bx128x17x17
+        :param torch.Tensor sent_embs: Sentence embeddings from text encoder. Bx256
+        :return: Logits for conditional adversarial loss. Bx1
+        :rtype: Any
+        """
+        return self.logits_cond(visual_features, sent_embs)
+
+    def get_uncond_logits(self, visual_features: torch.Tensor) -> Any:
```
Comment (Dmmc123) [inline src/models/modules/discriminator.py:144]: Wrapper functions `get_cond_logits`, `get_word_level_logits`, and `get_uncond_logits` are useless because you can just call logit extractors directly. They have the same notation even, no need to make a function that just calls one function with same params

File: src/models/modules/discriminator.py [hunk 3]
```diff
@@ -128,29 +128,50 @@ def __init__(self) -> None:
...
+        :param visual_features: Local features from Inception network. Bx128x17x17
+        :param word_embs: Word embeddings from text encoder. BxLx256
+        :return: Logits for word-level adversarial loss. BxL
+        :rtype: Any
+        """
+        return self.logits_word_level(visual_features, word_embs)
+
+    def forward(self, images: torch.Tensor) -> Any:
         """
         Obtain regional features for images and return logits
```
Comment (Dmmc123) [inline src/models/modules/discriminator.py:169]: You remade this in a way that `forward` no longer returns logits so in doc string you should write that it only retrieves features

File: src/models/modules/image_encoder.py [hunk 4]
```diff
@@ -112,14 +112,16 @@ def forward(self, image_tensor: torch.Tensor) -> Any:
 class VGGEncoder(nn.Module):
     """Pre Trained VGG Encoder Module"""
 
-    def __init__(self) -> None:
+    def __init__(self, device: Any) -> None:
```
Comment (Dmmc123) [inline src/models/modules/image_encoder.py:115]: `nn.Module` class is be default has a functionality for moving weights between devices. If another `nn.Module` is getting attached to base class through `self.` then when you apply `.to(device)` every weight part moves to that device. There's no reason to hardcode moving var `model`. You can just init it like `self.model = torch.hub.load.<bla-bla>` and you won't need to pass `device` argument

File: src/models/train_model.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        config_dict["output_dir"],
+        config_dict["snapshot"],
+        config_dict["const_dict"],
+    )
+
+    smooth_val_gen = const_dict["smooth_val_gen"]
+    lambda4 = const_dict["lambda4"]
+    generator = Generator(Ng, D, condition_dim, noise_dim).to(device)
+    discriminator = Discriminator().to(device)
+    text_encoder = TextEncoder(vocab_len, D, D // 2).to(device)
```
Comment (Dmmc123) [inline src/models/train_model.py:67]: is text encoder getting optimized? we're not using a pretrained encoder so you should initialize optimizer for it as well and maybe make it a part of discriminator or generator weights

File: src/models/train_model.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+                sent_emb,
+                labels_match,
+                correct_capt_len,
+                curr_class,
+                vgg_feat,
+                vgg_feat_fake,
+                const_dict,
+            )
+
+            loss_kl = kl_loss(mu_tensor, logvar)
```
Comment (Dmmc123) [inline src/models/train_model.py:175]: why `kl_loss` is not a part of `generator_loss` function if it gets added together with `loss_gen` ?

File: train.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        num_workers=args.num_workers,
+        multiprocessing_context = "spawn",
+        collate_fn=partial(custom_collate, device = cfg_dict["device"]),
+    )
+
+    # Train the model
+    train(dataloader, cfg_dict)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```
Comment (Dmmc123) [inline train.py:46]: kudos for using argparse!!!!

**Review by Dmmc123 — CHANGES_REQUESTED** [review]

File: .pylintrc (hunk 1 above)
Comment (rohit901) [reply .pylintrc:8]: it was giving me errors when i was calling text encoder, generator, discriminator, basically any pytorch module. Pytorch modules are callable but idk why it was showing that error, tried to Google it but didn't find any solution, if its possible to fix it without bypass can do it, let me know how.

File: src/models/modules/discriminator.py (hunk 2 above)
Comment (rohit901) [reply src/models/modules/discriminator.py:144]: chaned this, review and resolve please.

File: src/models/modules/discriminator.py (hunk 3 above)
Comment (rohit901) [reply src/models/modules/discriminator.py:169]: changed this.

File: src/models/modules/image_encoder.py (hunk 4 above)
Comment (rohit901) [reply src/models/modules/image_encoder.py:115]: changed this, thanks for telling me this. Didn't know about it.

File: src/models/train_model.py (hunk 5 above)
Comment (rohit901) [reply src/models/train_model.py:67]: ohhh you're right... should have done that. Thanks for catching.

File: train.py (hunk 7 above)
Comment (rohit901) [reply train.py:46]: hahah lol, i didn't know there is any alternative to argparse? like all the code I see uses only argparse.

**Review by Dmmc123 — APPROVED** [review]
